pinot-common/src/main/java/com/linkedin/pinot/common/utils/DocIdRange.java

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.linkedin.pinot.common.utils;
+
+import com.google.common.base.Preconditions;
+import java.util.Comparator;
+
+
+/**
+ * A document id range, which is inclusive on both ends. For example, a range of [1,3] would include documents 1, 2 and
+ * 3, but not 0 or 4.
+ */
+public class DocIdRange {
+  private static final boolean ENABLE_PRECONDITION_CHECKS = false;
+  int start;
+  int end;
+
+  public DocIdRange(int start, int end) {
+    this.start = start;
+    this.end = end;
+  }
+
+  /**
+   * Clips a range, making sure that its start value is greater or equals to the start of the clipping range and that
+   * its end value is smaller or equals to the end of the clipping range.
+   *
+   * @param clipRange The range to clip this range with
+   */
+  public void clip(DocIdRange clipRange) {
+    start = Math.max(start, clipRange.start);
+    end = Math.min(end, clipRange.end);
+  }
+
+  /**
+   * Checks whether this range is invalid (upper bound lower to the lower bound)
+   *
+   * @return true if the range is invalid
+   */
+  public boolean isInvalid() {
+    return end < start;
+  }
+
+  /**
+   * Checks whether ranges are mergeable (either overlapping or adjacent), assuming neither is invalid.
+   *
+   * @param otherRange The range to check for mergeability
+   * @return true if the given range can be merged
+   */
+  public boolean rangeIsMergeable(DocIdRange otherRange) {
+    if (ENABLE_PRECONDITION_CHECKS) {
+      Preconditions.checkArgument(!isInvalid());
+      Preconditions.checkArgument(!otherRange.isInvalid());
+    }
+
+    final boolean rangesAreAtLeastOneUnitApart = end < otherRange.start - 1 || otherRange.end < start - 1;
```

Comment:
this should check 
! otherRange.isInvalid()

Invalid otherRange can make this comparison incorrect since it's || condition